File: 백준/Gold/1516. 게임 개발/게임 개발.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());

        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        int[] indegree = new int[n + 1];
        int[] buildTime = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            buildTime[i] = Integer.parseInt(st.nextToken());

            while (true) {
                int prerequisite = Integer.parseInt(st.nextToken());
                if (prerequisite == -1) break;

                graph.get(prerequisite).add(i);
                indegree[i]++;
            }
        }

        Queue<Integer> queue = new LinkedList<>();
        int[] totalTime = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
                totalTime[i] = buildTime[i];
            }
        }

        while (!queue.isEmpty()) {
            int cur = queue.poll();
            for (int next : graph.get(cur)) {
                indegree[next]--;
                totalTime[next] = Math.max(totalTime[next], totalTime[cur] + buildTime[next]);

                if (indegree[next] == 0) queue.offer(next);
            }
        }

        for (int i = 1; i <= n; i++) {
            sb.append(totalTime[i]).append('\n');
        }

        System.out.print(sb);
    }
}

File: 백준/Gold/1516. 게임 개발/게임 개발.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList
import java.util.Queue

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val sb = StringBuilder()
    val n = br.readLine().toInt()

    val graph = List(n + 1) { mutableListOf<Int>() }
    val indegree = IntArray(n + 1)
    val buildTime = IntArray(n + 1)

    for (i in 1..n) {
        val tokens = br.readLine().split(" ").map { it.toInt() }
        buildTime[i] = tokens[0]

        for (j in 1 until tokens.size - 1) {
            val prerequisite = tokens[j]
            graph[prerequisite].add(i)
            indegree[i]++
        }
    }

    val queue: Queue<Int> = LinkedList()
    val totalTime = IntArray(n + 1)
    for (i in 1..n) {
        if (indegree[i] == 0) {
            queue.offer(i)
            totalTime[i] = buildTime[i]
        }
    }

    while (queue.isNotEmpty()) {
        val cur = queue.poll()
        for (next in graph[cur]) {
            indegree[next]--
            totalTime[next] = maxOf(totalTime[next], totalTime[cur] + buildTime[next])

            if (indegree[next] == 0) queue.offer(next)
        }
    }

    for (i in 1..n) {
        sb.append(totalTime[i]).append('\n')
    }

    print(sb)
}
